pkg/pillar/cas/cas.go

```diff
@@ -0,0 +1,89 @@
+package cas
+
+import (
+	"context"
+	"fmt"
+	"io"
+)
+
+//CAS  provides methods to interact with CAS clients
+type CAS interface {
+	//Blob APIs
+	//CheckBlobExists: returns true if the blob exists.
+	//Arg 'blobHash' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	CheckBlobExists(blobHash string) bool
+	//GetBlobSize: returns blob size.
+	//Arg 'blobHash' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	//Returns error if no blob is found for the given 'blobHash'.
+	GetBlobSize(blobHash string) (int64, error)
+	//ListBlobs: returns list of blobs where each entry is of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	ListBlobs() ([]string, error)
+	//IngestBlob: reads the blob as raw data from arg 'reader' and validates the read blob's hash with the provided arg 'blobHash'.
+	// If the validation succeeds, then the blob is ingested, else the given blob is rejected and an error is returned.
+	//Arg 'blobHash' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	//Arg 'ctx' must contain context with 'contextID':string with which we can relate two different blobs and
+	// 'expires':time.Duration to set a validity duration for the context
+	IngestBlob(ctx context.Context, blobHash string, reader io.Reader) error
+	//ReadBlob: returns a reader to consume the raw data of the blob which matches the given arg 'blobHash'.
+	//Returns error if no blob is found for the given 'blobHash'.
+	//Arg 'blobHash' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	ReadBlob(blobHash string) (io.Reader, error)
+	//RemoveBlob: removes a blob which matches the given arg 'blobHash'.
+	//To keep this method idempotent, no error is returned if the given arg 'blobHash' does not match any blob.
+	//Arg 'blobHash' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	RemoveBlob(blobHash string) error
+	//Children: returns a list of child blob hashes if the given arg 'blobHash' belongs to a
+	// index or a manifest blob, else an empty list is returned.
+	//Format of returned blob hash list and arg 'blobHash' is <algo>:<hash> (currently supporting only sha256:<hash>)
+	Children(blobHash string) ([]string, error)
+
+	//Image APIs
+	//CreateImage: creates a reference which points to a blob with 'blobHash'. 'blobHash' must belong to a index blob
```

Comment:
So this goes beyond just CAS and into Image and snapshot APIs @deitch  @adarsh-zededa . In principle I have no issue with it being merged as-is (and being refactored later), but it needs to be reconciled with:
   1. what we currently have under containerd/containerd
   2. the very idea that this interfaces is called CAS

Is there any reason to actually mix CAS, Images and Snapshots in a single interface?

Reply:
I think you are saying that we have 3 distinct activities:

* managing blobs of immutable content - uploading them, retrieving them, checking if they exist, etc.
* creating snapshots/volumes out of those blobs
* managing the relationship between image (just a tag/label) and the root blob of a content tree

I agree that these are distinct activities, but the only real, deployed, functional CAS systems that I know of - OCI registries and containerd - merge them. They merge them to the point of needing to integrate them because of garbage collection (blobs without images or parents or special tags are collected) and dependencies (need a blob to create a snapshot, etc.).

So I am not quite sure of any way to tease them apart at this point.